### PR TITLE
Patch 4

### LINKED ISSUE #32: fn
1. [closures.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/fn/closures.md)
2. [diverging.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/fn/diverging.md)

### PR DESCRIPTION
**Pull request by phongnguyen2012**
Closes: #32 #19 

update ngày 17/3/2023

### CODE REVIEW

**Comment by phongnguyen2012** [comment]
ngày 17/03/2023

**Review by phongnguyen2012 — COMMENTED** [review]
17/03/2023

File: src/fn/closures.md
Comment (HongThaiPham) [inline src/fn/closures.md:1]: file này sao nội dung khác vậy bạn ơi
https://github.com/rust-lang/rust-by-example/blob/master/src/fn/closures.md

**Comment by nghuyenthevinh2000** [comment]
1. https://github.com/rust-lang/rust-by-example/blob/master/src/fn/diverging.md
2.  https://github.com/rust-lang/rust-by-example/blob/master/src/fn/closures.md
3. https://github.com/rust-lang/rust-by-example/blob/master/src/error/result/result_alias.md
4. https://github.com/rust-lang/rust-by-example/blob/master/src/error/result/result_map.md

File: src/fn/diverging.md [hunk 1]
```diff
@@ -1 +1,41 @@
 # Diverging functions
+// Trong Rust, diverging funtion thường được sử dụng trong trường hợp xử lý lỗi.
```
Comment (nghuyenthevinh2000) [inline src/fn/diverging.md]:
```suggestion
// Trong Rust, diverging function thường được sử dụng trong trường hợp xử lý lỗi. 
```

File: src/fn/diverging.md [hunk 2]
```diff
@@ -1 +1,41 @@
 # Diverging functions
+// Trong Rust, diverging funtion thường được sử dụng trong trường hợp xử lý lỗi. 
+// Khi một lỗi không thể được xử lý và chương trình không thể tiếp tục thực hiện, 
+//thì các diverging function được sử dụng để kích hoạt lỗi và dừng chương trình
+
+// Hàm foo() không có giá trị trả về nhưng thay vào đó, nó kích hoạt một sử dụng panic!() để dừng chương trình.
```
Comment (nghuyenthevinh2000) [inline src/fn/diverging.md]:
```suggestion
// Hàm foo() không có giá trị trả về nhưng thay vào đó, nó kích hoạt hàm panic!() để dừng chương trình. 
```

**Review by nghuyenthevinh2000 — CHANGES_REQUESTED** [review]

**Comment by phongnguyen2012** [comment]
Closes: #19 #32 
[result_alias.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/error/result/result_alias.md)
[result_map.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/error/result/result_map.md)
[closures.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/fn/closures.md)
[diverging.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/fn/diverging.md)

File: src/scope/lifetime/methods.md
Comment (duonghb53) [inline src/scope/lifetime/methods.md:1]: Anh @phongnguyen2012 thêm lại 2 file này như cũ nhé

File: src/scope/lifetime/static_lifetime.md
Comment (duonghb53) [inline src/scope/lifetime/static_lifetime.md:1]: Anh @phongnguyen2012 thêm lại 2 file này như cũ nhé

File: src/error/result/result_alias.md [hunk 3]
```diff
@@ -1 +1,44 @@
-# aliases for Result
+# aliases for `Result`
+Khi chúng ta muốn sử dụng lại một kiểu Result cụ thể nhiều lần thì làm sao?
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
Khi chúng ta muốn tái sử dụng một kiểu `Result` cụ thể nhiều lần thì làm như thế nào?
```

File: src/error/result/result_alias.md [hunk 4]
```diff
@@ -1 +1,44 @@
...
+Hãy nhớ rằng Rust cho phép chúng ta tạo [aliases][typealias]. Conveniently,
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
Hãy nhớ rằng Rust cho phép chúng ta tạo [aliases][typealias]. 
```

File: src/error/result/result_alias.md [hunk 5]
```diff
@@ -1 +1,44 @@
...
+Một cách tiện lợi, chúng ta có thể định nghĩa một bí danh cho kiểu Result cụ thể đó.
+
+Ở mức độ module, việc tạo bí danh có thể rất hữu ích. Các lỗi được tìm thấy trong một module cụ thể thường có cùng kiểu Err,
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
Ở mức độ module, việc tạo bí danh có thể rất hữu ích. Các lỗi được tìm thấy trong một module cụ thể thường có cùng kiểu `Err`,
```

File: src/error/result/result_alias.md [hunk 6]
```diff
@@ -1 +1,44 @@
...
+do đó một bí danh duy nhất có thể ngắn gọn định nghĩa tất cả các Result liên quan.
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
do đó một bí danh duy nhất có thể ngắn gọn định nghĩa tất cả các `Result` liên quan. 
```

File: src/error/result/result_alias.md [hunk 7]
```diff
@@ -1 +1,44 @@
-# aliases for Result
+# aliases for `Result`
+Khi chúng ta muốn sử dụng lại một kiểu Result cụ thể nhiều lần thì làm sao?
+Hãy nhớ rằng Rust cho phép chúng ta tạo [aliases][typealias]. Conveniently,
+Một cách tiện lợi, chúng ta có thể định nghĩa một bí danh cho kiểu Result cụ thể đó.
+
+Ở mức độ module, việc tạo bí danh có thể rất hữu ích. Các lỗi được tìm thấy trong một module cụ thể thường có cùng kiểu Err,
+do đó một bí danh duy nhất có thể ngắn gọn định nghĩa tất cả các Result liên quan. 
+Việc này rất hữu ích đến mức thư viện std cung cấp sẵn một bí danh cho việc này: [`io::Result`][io_result]!
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
Việc này rất hữu ích đến mức thư viện `std` cung cấp sẵn một bí danh cho việc này: [`io::Result`][io_result]!
```

File: src/error/result/result_alias.md [hunk 8]
```diff
@@ -1 +1,44 @@
...
+
+Đây là một ví dụ nhanh để minh họa cú pháp:
+
+```rust,editable
+use std::num::ParseIntError;
+
+// Định nghĩa một bí danh chung cho Result với kiểu lỗi ParseIntError.
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
// Định nghĩa một bí danh chung cho `Result` với kiểu lỗi `ParseIntError`.
```

File: src/error/result/result_alias.md [hunk 9]
```diff
@@ -1 +1,44 @@
...
+type AliasedResult<T> = Result<T, ParseIntError>;
+
+// Sử dụng bí danh trên để tham chiếu đến kiểu Result cụ thể của chúng ta.
```
Comment (duonghb53) [inline src/error/result/result_alias.md]:
```suggestion
// Sử dụng bí danh trên để tham chiếu đến kiểu `Result` cụ thể của chúng ta.
```

File: src/error/result/result_map.md [hunk 10]
```diff
@@ -1 +1,83 @@
-# map for Result
+# `map` for `Result`
+
+Khi xảy ra lỗi trong hàm `multiply` của ví dụ trước, sử dụng panic không tạo ra mã nguồn vững chắc.
```
Comment (duonghb53) [inline src/error/result/result_map.md]:
```suggestion
Khi xảy ra lỗi trong hàm `multiply` của ví dụ trước, sử dụng panic không tạo ra mã nguồn mạnh mẽ. 
```

File: src/error/result/result_map.md [hunk 11]
```diff
@@ -1 +1,83 @@
-# map for Result
+# `map` for `Result`
+
+Khi xảy ra lỗi trong hàm `multiply` của ví dụ trước, sử dụng panic không tạo ra mã nguồn vững chắc. 
+Thông thường, chúng ta muốn trả lại lỗi cho người gọi hàm để người gọi hàm quyết định cách phản hồi với lỗi đó.
+Đầu tiên, chúng ta cần phải biết loại lỗi mà chúng ta đang gặp phải. Để xác định kiểu Err,
+chúng ta tìm kiểu lỗi trong: [`parse()`][parse], Được thực hiện bằng cách sử dụng
```
Comment (duonghb53) [inline src/error/result/result_map.md]:
```suggestion
chúng ta tìm kiểu lỗi trong [`parse()`][parse], Được thực hiện bằng cách triển khai trait
```

File: src/error/result/result_map.md [hunk 12]
```diff
@@ -1 +1,83 @@
...
+[`FromStr`][from_str] trait for [`i32`][i32]. Kết quả, kiểu `Err` được xác định như sau: [`ParseIntError`][parse_int_error].
```
Comment (duonghb53) [inline src/error/result/result_map.md]:
```suggestion
[`FromStr`][from_str] cho [`i32`][i32]. Kết quả là, kiểu `Err` được xác định là kiểu [`ParseIntError`][parse_int_error].
```

File: src/error/result/result_map.md [hunk 13]
```diff
@@ -1 +1,83 @@
...
+    let twenty = multiply("10", "2");
+    print(twenty);
+
+    // Bây giờ đoạn mã sau đây cung cấp một thông báo lỗi hữu ích hơn nhiều.
+    let tt = multiply("t", "2");
+    print(tt);
+}
+```
+May mắn thay, các phương thức `Option`'s `map`, `and_then` và nhiều phương thức khác cũng được cài đặt cho Result.
+[`Result`][result] chứa một danh sách đầy đủ..
```
Comment (duonghb53) [inline src/error/result/result_map.md]:
```suggestion
[`Result`][result] chứa một danh sách đầy đủ.
```

File: src/error/result/result_map.md [hunk 14]
```diff
@@ -1 +1,83 @@
...
+   // Điều này vẫn cho phép trả về một câu trả lời hợp lý.
+    let twenty = multiply("10", "2");
+    print(twenty);
+
+    // Bây giờ đoạn mã sau đây cung cấp một thông báo lỗi hữu ích hơn nhiều.
+    let tt = multiply("t", "2");
+    print(tt);
+}
+```
+May mắn thay, các phương thức `Option`'s `map`, `and_then` và nhiều phương thức khác cũng được cài đặt cho Result.
```
Comment (duonghb53) [inline src/error/result/result_map.md]:
```suggestion
May mắn thay, các phương thức `Option`'s `map`, `and_then` và nhiều phương thức khác cũng được cài đặt cho `Result`.
```

File: src/error/result/result_map.md [hunk 15]
```diff
@@ -1 +1,83 @@
...
+[`Result`][result] chứa một danh sách đầy đủ..
+
+```rust,editable
+use std::num::ParseIntError;
+
+// Giống với `Option`, chúng ta có thể sử dụng các phương thức ghép kết quả như `map()`.
```
Comment (duonghb53) [inline src/error/result/result_map.md]:
```suggestion
// Giống với `Option`, chúng ta có thể sử dụng các phương thức kết hợp như `map()`.
```

File: src/fn/closures.md [hunk 16]
```diff
@@ -1 +1,47 @@
 # Closures
+
+Closures là các hàm có thể nắm giữ môi trường bao quanh. Ví dụ, closure nắm giữ biến `x`:
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
Closures là các hàm có thể capture môi trường bao quanh. Ví dụ, closure capture biến `x`:
```

File: src/fn/closures.md [hunk 17]
```diff
@@ -1 +1,47 @@
...
+
+```Rust
+|val| val + x
+```
+
+Cú pháp và khả năng của closures làm cho chúng rất tiện lợi cho việc sử dụng ngay lập tức.
+Gọi một closure hoàn toàn giống như gọi một hàm..
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
Gọi một closure hoàn toàn giống như gọi một hàm.
```

File: src/fn/closures.md [hunk 18]
```diff
@@ -1 +1,47 @@
...
+Tuy nhiên, cả kiểu dữ liệu đầu vào và đầu ra *can* được suy luận và tên biến đầu vào *must* được chỉ định.
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
Tuy nhiên, cả kiểu dữ liệu đầu vào và đầu ra có thể được suy luận và tên biến đầu vào phải được chỉ định.
```

File: src/fn/closures.md [hunk 19]
```diff
@@ -1 +1,47 @@
...
+
+Một số đặc điểm khác của closures bao gồm:
+* Sử dụng `||` thay vì `()` để bao quanh biến đầu vào..
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
* Sử dụng `||` thay vì `()` để bao quanh biến đầu vào.
```

File: src/fn/closures.md [hunk 20]
```diff
@@ -1 +1,47 @@
...
+* Tùy chọn việc đặt dấu mở ngoặc nhọn (`{}`) cho một biểu thức đơn (bắt buộc nếu không phải là biểu thức đơn).
+* Khả năng bắt được các biến môi trường bên ngoài (outer environment variables).
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
* Khả năng capture được các biến môi trường bên ngoài (outer environment variables).
```

File: src/fn/closures.md [hunk 21]
```diff
@@ -1 +1,47 @@
...
+    
+    // Một hàm thông thường không thể tham chiếu đến các biến trong môi trường bao quanh nó.
+    //fn function(i: i32) -> i32 { i + outer_var }
+    // TODO:  xóa comment trên và xem lỗi biên dịch. Trình biên dịch
+    // gợi ý chúng ta nên định nghĩa một closure thay vì.
+
+    // Closures là vô danh, ở đây chúng ta đang liên kết chúng với các tham chiếu.
+    // Chú thích giống với chú thích của hàm, nhưng là tùy chọn
+    // giống như các dấu ngoặc nhọn `{}` bao quanh thân hàm. Những hàm vô danh này
+    // được gán cho các biến được đặt tên phù hợp..
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
    // được gán cho các biến được đặt tên phù hợp.
```

File: src/fn/closures.md [hunk 22]
```diff
@@ -1 +1,47 @@
...
+    let closure_annotated = |i: i32| -> i32 { i + outer_var };
+    let closure_inferred  = |i     |          i + outer_var  ;
+
+    // Gọi các closures.
+    println!("closure_annotated: {}", closure_annotated(1));
+    println!("closure_inferred: {}", closure_inferred(1));
+    // Một khi kiểu dữ liệu của closure đã được suy luận, thì nó không thể được suy luận lại với một kiểu dữ liệu khác.
```
Comment (duonghb53) [inline src/fn/closures.md]:
```suggestion
    // Một khi kiểu dữ liệu của closure đã được suy ra, thì nó không thể được suy ra lại với một kiểu dữ liệu khác.
```

File: src/scope/lifetime/methods.md
Comment (duonghb53) [reply src/scope/lifetime/methods.md:1]: @phongnguyen2012 em thấy a resolved nhưng chưa add file?

File: src/fn/closures.md (hunk 18 above)
Comment (duonghb53) [reply src/fn/closures.md]:
```suggestion
Tuy nhiên, cả kiểu dữ liệu đầu vào và đầu ra có thể được suy luận và tên biến đầu vào phải được chỉ định.
```

**Review by duonghb53 — APPROVED** [review]
LGTM

**Review by HongThaiPham — APPROVED** [review]

**Review by nghuyenthevinh2000 — APPROVED** [review]

**Pull request merged**